File: src/whylabs/logs/core/columnprofile.py
#!/usr/bin/env python3
"""
created 5/7/20 by ibackus
"""
from whylabs.logs.core.types import TypedDataConverter
from whylabs.logs.core.statistics import NumberTracker
from whylabs.logs.core.statistics import from_number_tracker
from whylabs.logs.proto import ColumnSummary, ColumnMessage, InferredType
from whylabs.logs.core.statistics import CountersTracker, SchemaTracker
from whylabs.logs.core.statistics.datatypes import StringTracker
_TYPES = InferredType.Type
_NUMERIC_TYPES = {_TYPES.FRACTIONAL, _TYPES.INTEGRAL}


class ColumnProfile:
    """
    Statistics tracking for a column

    Parameters
    ----------
    name : str (required)
        Name of the column profile
    number_tracker : NumberTracker
        Implements numeric data statisics tracking
    string_tracker : StringTracker
        Implements string data-type statistics tracking
    schema_tracker : SchemaTracker
        Implements tracking of schema-related information
    counters : CountersTracker
        Keep count of various things

    TODO:
        * Proper TypedDataConverter type checking
        * Multi-threading/parallelism
    """

    def __init__(self, name, number_tracker=None, string_tracker=None,
                 schema_tracker=None, counters=None):
        # Handle default values
        if counters is None:
            counters = CountersTracker()
        if number_tracker is None:
            number_tracker = NumberTracker()
        if string_tracker is None:
            string_tracker = StringTracker()
        if schema_tracker is None:
            schema_tracker = SchemaTracker()
        # Assign values
        self.column_name = name
        self.number_tracker = number_tracker
        self.string_tracker = string_tracker
        self.schema_tracker = schema_tracker
        self.counters = counters

    def track(self, value):
        """
        Add `value` to tracking statistics.
        """
        self.counters.increment_count()
        if value is None:
            self.counters.increment_null()
            return

        # TODO: ignore this if we already know the data type
        if isinstance(value, str):
            self.string_tracker.update(value)

        # TODO: Implement real typed data conversion
        typed_data = TypedDataConverter.convert(value)
        dtype = TypedDataConverter.get_type(typed_data)
        self.schema_tracker.track(dtype)

        if isinstance(typed_data, bool):
            # Note: bools are sub-classes of ints in python, so we should check
            # for bool type first
            self.counters.increment_bool()
        elif isinstance(typed_data, (float, int)):
            self.number_tracker.track(typed_data)

    def to_summary(self):
        """
        Generate a summary of the statistics

        Returns
        -------
        summary : ColumnSummary
            Protobuf summary message.
        """
        schema = None
        if self.schema_tracker is not None:
            schema = self.schema_tracker.to_summary()
        # TODO: implement the real schema/type checking
        opts = dict(
            counters=self.counters.to_protobuf()
        )
        if schema is not None:
            opts['schema'] = schema
            dtype = schema.inferred_type.type
            if dtype == _TYPES.STRING:
                if self.string_tracker is not None:
                    string_summary = self.string_tracker.to_summary()
                    opts['string_summary'] = string_summary

            elif dtype in _NUMERIC_TYPES:
                numbers_summary = from_number_tracker(self.number_tracker)
                if numbers_summary is not None:
                    opts['number_summary'] = numbers_summary
        return ColumnSummary(**opts)

    def merge(self, other):
        """
        Merge this columnprofile with another.

        Parameters
        ----------
        other : ColumnProfile

        Returns
        -------
        merged : ColumnProfile
            A new, merged column profile.
        """
        assert self.column_name == other.column_name
        return ColumnProfile(
            self.column_name,
            number_tracker=self.number_tracker.merge(other.number_tracker),
            string_tracker=self.string_tracker.merge(other.string_tracker),
            schema_tracker=self.schema_tracker.merge(other.schema_tracker),
            counters=self.counters.merge(other.counters),
        )

    def to_protobuf(self):
        """
        Return the object serialized as a protobuf message

        Returns
        -------
        message : ColumnMessage
        """
        return ColumnMessage(
            name=self.column_name,
            counters=self.counters.to_protobuf(),
            schema=self.schema_tracker.to_protobuf(),
            numbers=self.number_tracker.to_protobuf(),
            strings=self.string_tracker.to_protobuf(),
        )

    @staticmethod
    def from_protobuf(message):
        """
        Load from a protobuf message

        Returns
        -------
        column_profile : ColumnProfile
        """
        return ColumnProfile(
            message.name,
            counters=CountersTracker.from_protobuf(message.counters),
            schema_tracker=SchemaTracker.from_protobuf(message.schema),
            number_tracker=NumberTracker.from_protobuf(message.numbers),
            string_tracker=StringTracker.from_protobuf(message.strings),
        )

File: core/src/main/java/com/whylogs/core/ColumnProfile.java
package com.whylogs.core;

import static com.whylogs.core.SummaryConverters.fromSchemaTracker;
import static com.whylogs.core.statistics.datatypes.StringTracker.ARRAY_OF_STRINGS_SER_DE;
import static com.whylogs.core.types.TypedDataConverter.NUMERIC_TYPES;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Sets;
import com.google.protobuf.ByteString;
import com.whylogs.core.message.ColumnMessage;
import com.whylogs.core.message.ColumnSummary;
import com.whylogs.core.message.HllSketchMessage;
import com.whylogs.core.message.InferredType;
import com.whylogs.core.statistics.CountersTracker;
import com.whylogs.core.statistics.NumberTracker;
import com.whylogs.core.statistics.SchemaTracker;
import com.whylogs.core.statistics.datatypes.StringTracker;
import com.whylogs.core.types.TypedData;
import com.whylogs.core.types.TypedDataConverter;
import com.whylogs.core.utils.sketches.FrequentStringsSketch;
import javax.annotation.Nullable;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.val;
import org.apache.datasketches.frequencies.ItemsSketch;
import org.apache.datasketches.hll.HllSketch;
import org.apache.datasketches.hll.Union;
import org.apache.datasketches.memory.Memory;

@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Getter
@Builder(setterPrefix = "set")
public class ColumnProfile {
  public static final int FREQUENT_MAX_LG_K = 7;
  private static final int CARDINALITY_LG_K = 12;
  public static final int STRING_LENGTH_MAX = 256;
  private static volatile ImmutableSet<String> NULL_STR_ENVS;

  @NonNull private final String columnName;
  @NonNull private final CountersTracker counters;
  @NonNull private final SchemaTracker schemaTracker;
  @NonNull private final NumberTracker numberTracker;
  @NonNull private final ItemsSketch<String> frequentItems;
  @NonNull private final HllSketch cardinalityTracker;
  @NonNull private final ImmutableSet<String> nullStrs;
  @NonNull private final StringTracker stringTracker;

  static ImmutableSet<String> nullStrsFromEnv() {
    if (ColumnProfile.NULL_STR_ENVS == null) {
      val nullSpec = System.getenv("NULL_STRINGS");
      ColumnProfile.NULL_STR_ENVS =
          nullSpec == null ? ImmutableSet.of() : ImmutableSet.copyOf(nullSpec.split(","));
    }
    return ColumnProfile.NULL_STR_ENVS;
  }

  public ColumnProfile(String columnName) {
    this(columnName, nullStrsFromEnv());
  }

  public ColumnProfile(String columnName, ImmutableSet<String> nullStrs) {
    this.columnName = columnName;
    this.counters = new CountersTracker();
    this.schemaTracker = new SchemaTracker();
    this.stringTracker = new StringTracker();
    this.numberTracker = new NumberTracker();
    this.frequentItems = FrequentStringsSketch.create();
    this.cardinalityTracker = new HllSketch(CARDINALITY_LG_K);
    this.nullStrs = nullStrs;
  }

  public void track(Object value) {
    synchronized (this) {
      counters.incrementCount();

      // TODO: ignore this if we already know the data type
      val typedData = TypedDataConverter.convert(value);
      if (isNull(typedData)) {
        schemaTracker.track(InferredType.Type.NULL);
        return;
      }
      schemaTracker.track(typedData.getType());

      switch (typedData.getType()) {
        case FRACTIONAL:
          final double fractional = typedData.getFractional();
          trackText(String.valueOf(fractional));
          numberTracker.track(fractional);
          break;
        case INTEGRAL:
          final long integralValue = typedData.getIntegralValue();
          trackText(String.valueOf(integralValue));
          numberTracker.track(integralValue);
          break;
        case BOOLEAN:
          // TODO: handle boolean across languages? Python booleans are "True" vs Java "true"
          trackText(String.valueOf(typedData.isBooleanValue()));
          if (typedData.isBooleanValue()) {
            counters.incrementTrue();
          }
          break;
        case STRING:
          final String stringValue = typedData.getStringValue();

          // trackText is the older tracking code.  It limits the length of strings in order to
          // track cardinality.
          trackText(stringValue);

          // string tracking does not limit the length of input text.
          stringTracker.update(stringValue);
      }
    }
  }

  private boolean isNull(TypedData value) {
    if (value == null) return true;
    if (value.getType() == InferredType.Type.STRING && !this.nullStrs.isEmpty()) {
      return this.nullStrs.contains(value.getStringValue());
    }

    if (value.getType() == InferredType.Type.FRACTIONAL) {
      return Double.isNaN(value.getFractional()) || Double.isInfinite(value.getFractional());
    }

    return false;
  }

  private void trackText(String text) {
    if (text != null && text.length() > STRING_LENGTH_MAX) {
      text = text.substring(0, STRING_LENGTH_MAX);
    }
    frequentItems.update(text);
    cardinalityTracker.update(text);
  }

  public ColumnSummary toColumnSummary() {
    val schema = fromSchemaTracker(schemaTracker);

    val builder = ColumnSummary.newBuilder().setCounters(counters.toProtobuf());
    if (schema != null) {
      builder.setSchema(schema);
      if (NUMERIC_TYPES.contains(schema.getInferredType().getType())) {
        val stringsSummary = SummaryConverters.fromStringTracker(this.stringTracker);

        val numberSummary = SummaryConverters.fromNumberTracker(this.numberTracker);
        if (numberSummary != null) {
          builder.setNumberSummary(numberSummary);
        }
      }
    }
    return builder.build();
  }

  public ColumnProfile merge(ColumnProfile other) {
    return this.merge(other, true);
  }

  public ColumnProfile merge(ColumnProfile other, boolean checkName) {
    if (checkName) {
      Preconditions.checkArgument(
          this.columnName.equals(other.columnName),
          String.format(
              "Mismatched column name. Expected [%s], got [%s]",
              this.columnName, other.columnName));
    }

    val mergedSketch = Union.heapify(this.cardinalityTracker.toCompactByteArray());
    mergedSketch.update(other.cardinalityTracker);

    val iMem = Memory.wrap(frequentItems.toByteArray(ARRAY_OF_STRINGS_SER_DE));
    val copyFreqItems = ItemsSketch.getInstance(iMem, ARRAY_OF_STRINGS_SER_DE);
    copyFreqItems.merge(other.frequentItems);

    val builder =
        ColumnProfile.builder()
            .setColumnName(this.columnName)
            .setCounters(this.counters.merge(other.counters))
            .setNumberTracker(this.numberTracker.merge(other.numberTracker))
            .setSchemaTracker(this.schemaTracker.merge(other.schemaTracker))
            .setCardinalityTracker(HllSketch.heapify(mergedSketch.toCompactByteArray()))
            .setFrequentItems(copyFreqItems)
            .setNullStrs(Sets.union(this.getNullStrs(), other.getNullStrs()).immutableCopy());

    // backward compatibility with profiles that don't have stringtracker.
    if (this.stringTracker == null) {
      builder.setStringTracker(other.stringTracker);
    } else {
      builder.setStringTracker(this.stringTracker.merge(other.stringTracker));
    }

    return builder.build();
  }

  public ColumnMessage.Builder toProtobuf() {
    val hllSketchMessage =
        HllSketchMessage.newBuilder()
            .setLgK(cardinalityTracker.getLgConfigK())
            .setSketch(ByteString.copyFrom(cardinalityTracker.toCompactByteArray()));

    return ColumnMessage.newBuilder()
        .setName(columnName)
        .setCounters(counters.toProtobuf())
        .setSchema(schemaTracker.toProtobuf())
        .setNumbers(numberTracker.toProtobuf())
        .setStrings(stringTracker.toProtobuf())
        .setCardinalityTracker(hllSketchMessage)
        .setFrequentItems(FrequentStringsSketch.toStringSketch(this.frequentItems));
  }

  @Nullable
  public static ColumnProfile fromProtobuf(@Nullable ColumnMessage message) {
    if (message == null || message.getSerializedSize() == 0) {
      return null;
    }

    val builder =
        ColumnProfile.builder()
            .setColumnName(message.getName())
            .setCounters(CountersTracker.fromProtobuf(message.getCounters()))
            .setSchemaTracker(
                SchemaTracker.fromProtobuf(
                    message.getSchema(), message.getCounters().getNullCount().getValue()))
            .setNumberTracker(NumberTracker.fromProtobuf(message.getNumbers()))
            .setCardinalityTracker(
                HllSketch.heapify(message.getCardinalityTracker().getSketch().toByteArray()))
            .setFrequentItems(
                FrequentStringsSketch.deserialize(message.getFrequentItems().getSketch()))
            .setNullStrs(ColumnProfile.nullStrsFromEnv());

    // backward compatibility - only decode StringsMessage if it exists.
    // older profiles written by java library may not have any StringsMessage.
    StringTracker strTracker;
    if (message.hasStrings()) {
      strTracker = StringTracker.fromProtobuf(message.getStrings());
    } else {
      strTracker = new StringTracker();
    }
    builder.setStringTracker(strTracker);

    return builder.build();
  }
}
